compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala

```diff
@@ -636,6 +647,232 @@ class TreePickler(pickler: TastyPickler) {
+        }
+      case Alternative(alts) =>
+        writeByte(ALTERNATIVE)
+        withLength { alts.foreach(pickleUntyped) }
+      case tree: untpd.ValDef =>
+        pickleDef(VALDEF, tree, tree.tpt, tree.rhs)
+      case tree: untpd.DefDef =>
+        pickleDef(DEFDEF, tree, tree.tpt, tree.rhs, pickleAllParams(tree))
+      case tree: untpd.TypeDef =>
+        pickleDef(TYPEDEF, tree, tree.rhs)
+      case tree: untpd.ModuleDef =>
+        pickleDef(OBJECTDEF, tree, tree.impl)
+      case tree: untpd.Template =>
+        writeByte(TEMPLATE)
+        tree.parents.foreach(pickleParent)
+        if (!tree.self.isEmpty) {
+          writeByte(SELFDEF); pickleName(tree.self.name); pickleTpt(tree.self.tpt)
+        }
+        pickleUntyped(tree.constr)
+        tree.body.foreach(pickleUntyped)
+      case Import(expr, selectors) =>
+        writeByte(IMPORT)
+        withLength { pickleUntyped(expr); pickleSelectors(selectors) }
+      case tree: untpd.TypeTree =>
+        pickleDummyType()
+      case SingletonTypeTree(ref) =>
+        writeByte(SINGLETONtpt)
+        pickleTerm(ref)
+      case RefinedTypeTree(parent, refinements) =>
+        writeByte(REFINEDtpt)
+        withLength { pickleTpt(parent); refinements.foreach(pickleTerm) }
+      case AppliedTypeTree(tycon, args) =>
+        writeByte(APPLIEDtpt)
+        withLength { pickleTpt(tycon); args.foreach(pickleTpt) }
+      case AndTypeTree(tp1, tp2) =>
+        writeByte(ANDtpt)
+        withLength { pickleTpt(tp1); pickleTpt(tp2) }
+      case OrTypeTree(tp1, tp2) =>
+        writeByte(ORtpt)
+        withLength { pickleTpt(tp1); pickleTpt(tp2) }
+      case ByNameTypeTree(tp) =>
+        writeByte(BYNAMEtpt)
+        pickleTpt(tp)
+      case Annotated(tree, annot) =>
+        writeByte(ANNOTATEDtpt)
+        withLength { pickleTpt(tree); pickleTerm(annot) }
+      case LambdaTypeTree(tparams, body) =>
+        writeByte(LAMBDAtpt)
+        withLength { pickleParams(tparams); pickleTpt(body) }
+      case TypeBoundsTree(lo, hi) =>
+        writeByte(TYPEBOUNDStpt)
+        withLength {
+          pickleTpt(lo);
+          if (hi ne lo) pickleTpt(hi)
+        }
+      case untpd.Function(args, body) =>
+        writeByte(FUNCTION)
+        withLength { pickleUntyped(body); args.foreach(pickleUntyped) }
+      case untpd.InfixOp(l, op, r) =>
+        writeByte(INFIXOP)
+        withLength { pickleUntyped(l); pickleUntyped(op); pickleUntyped(r) }
+      case untpd.PatDef(mods, pats, tpt, rhs) =>
+        writeByte(PATDEF)
+        withLength {
+          pickleTpt(tpt)
+          pickleUntyped(rhs)
+          pats.foreach(pickleUntyped)
+          pickleModifiers(mods, isTerm = true)
+        }
+      case untpd.TypedSplice(splice) =>
+        writeByte(TYPEDSPLICE)
+        withLength { pickleTree(splice) }
+      case Thicket(trees) =>
+        if (trees.isEmpty) writeByte(EMPTYTREE)
+        else trees.foreach(pickleUntyped)
```

Comment:
Shouldn't we have `withLength {trees.foreach(pickleUntyped)}` to then know how many trees we need to read?

Reply:
No, since non-empty thickets are inlined.